File: leetcode/string/l20.ts
function isValid(s: string): boolean {
    let stack = [];
    if (s.length == 1) return false;
    for (let i = 0; i < s.length; i++) {
        if (isLeft(s.charAt(i))) {
            stack.push(s.charAt(i));
        } else {
            if (stack.length == 0) return false;
            let left = stack.pop();
            if (!isMatch(left, s.charAt(i))) return false;
        }
    }
    return stack.length == 0;
};

function isMatch(left: string, right: string): boolean {
    let flag = true;
    switch (left) {
        case '(':
            flag = right == ')';
            break;
        case '{':
            flag = right == '}';
            break;
        case '[':
            flag = right == ']';
            break;
        default:
            flag = false;
    }
    return flag;
}

function isLeft(left: string): boolean {
    switch (left) {
        case '(':
        case '[':
        case '{':
            return true;
    }
    return false;
}

File: src/l20.cpp
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

class Solution {
public:
    std::unordered_set<char> left {
        '(', '{', '['
    };
    std::unordered_set<char> right {
        ')', '}', ']'
    };
    bool is_left(char ch) {
        return left.find(ch) != left.end();
    }
    bool is_right(char ch) {
        return right.find(ch) != right.end();
    }
    bool is_match(char ch_l, char ch_r) {
        if (ch_l == '(' && ch_r == ')') return true;
        if (ch_l == '[' && ch_r == ']') return true;
        if (ch_l == '{' && ch_r == '}') return true;
        return false;
    }
    bool isValid(std::string s) {
        std::vector<int> stack;
        bool result = true;
        for(auto ch : s) {
            if (is_left(ch)) {
                stack.push_back(ch);
            } else {
                if (stack.empty()) {
                    result = false;
                    break;
                }
                bool match = is_match(stack.back(),ch);
                if (!match) {
                    result = false;
                    break;
                }
                stack.pop_back();
            }
        }
        return result ? stack.empty() : result;
    }
};